sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -3237,6 +3238,28 @@ class Dataset[T] private[sql](
     files.toSet.toArray
   }
 
+  /**
+   * Returns the tuple of the row count and an SeqView that contains all rows in this Dataset.
+   *
+   * The SeqView will consume as much memory as the total size of serialized results which can be
+   * limited with the config 'spark.driver.maxResultSize'. Rows are deserialized when iterating rows
+   * with iterator of returned SeqView. Whether to collect all deserialized rows or to iterate them
+   * incrementally can be decided with considering total rows count and driver memory.
+   */
+  private[sql] def collectCountAndSeqView(): (Long, SeqView[T, Array[T]]) =
+    withAction("collectCountAndSeqView", queryExecution) { plan =>
+      // This projection writes output to a `InternalRow`, which means applying this projection is
+      // not thread-safe. Here we create the projection inside this method to make `Dataset`
+      // thread-safe.
+      val objProj = GenerateSafeProjection.generate(deserializer :: Nil)
+      val (totalRowCount, internalRowsView) = plan.executeCollectSeqView()
+      (totalRowCount, internalRowsView.map { row =>
+        // The row returned by SafeProjection is `SpecificInternalRow`, which ignore the data type
+        // parameter of its `get` method, so it's safe to use null here.
+        objProj(row).get(0, null).asInstanceOf[T]
+      }.asInstanceOf[SeqView[T, Array[T]]])
+    }
```

Comment:
If this is a thriftserver specific issue, can we do the same thing by fixing code only in the thriftserver package?
IMHO we'd be better not to modify code in the sql package as much as possible.

Reply:
Currently, there is no API to deserialize result iteratively in Dataset.

Reply:
Yes, that's what I mean. I thought that 'deserializer' is declared with private, so there is now way to get 'deserializer' out of Dataset.

Reply:
how about changing private to private[sql], then implementing this based on the deserializer?

Reply:
Ok, you don't prefer adding function to Dataset. If withAction & deserializer are changed to private[sql], this implementation can be moved out. Is this function useful for other SQL server to reduce memory usage of query execution? I don't think it looks good because Projection is created in the outside of Dataset.